src/components/org/apache/jmeter/control/WeightedDistributionController.java

```diff
@@ -0,0 +1,472 @@
+ * 
+ * The list of subcontrollers is generated differently if called during test execution rather
+ * Test plan building gui mode, this abstracts the way that are accessed
+ */
+class SubControllerIterator implements Iterator<TestElement> {
+    
+    @SuppressWarnings("rawtypes")
+    private Enumeration subControllerEnumeration = null;
+        
+    private TestElement nextSubController;
+    
+    public SubControllerIterator(WeightedDistributionController wdc) {
+        subControllerEnumeration = null;
+        nextSubController = null;
+        
+        // When calling in during test execution, the child elements are
+        // listed in getSubControllers, but during test design time,
+        // elements need to be pulled from the node tree
+        if (wdc.getSubControllers().size() > 0) {
+            subControllerEnumeration =  new Vector<TestElement>(wdc.getSubControllers()).elements();
+        } else if (wdc.getNode() != null
+                && wdc.getNode().children().hasMoreElements()) {
+            subControllerEnumeration =  wdc.getNode().children();
+        } else {
+            WeightedDistributionController.log.warn("Unable to find subcontrollers");
+            subControllerEnumeration = null;
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (nextSubController != null) {
+            return true;
+        } else {
+            setNextSubController();
+            return nextSubController != null;
+        }
+    }
+
+    @Override
+    public TestElement next() {
+        if (nextSubController == null) {
+            setNextSubController();
+            if (nextSubController == null) {
+                throw new NoSuchElementException();
+            }
+        }
+        
+        TestElement returnSubController = nextSubController;
+        nextSubController = null;
+        
+        return returnSubController;
+    }
+    
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("Iterator.remove() not implemented");
+    }
+    
+    private void setNextSubController() {
+        if (subControllerEnumeration != null && subControllerEnumeration.hasMoreElements()) {
+            Object currSubCtrl = subControllerEnumeration.nextElement();
+            
+            if (currSubCtrl instanceof JMeterTreeNode) {
+                nextSubController = ((JMeterTreeNode) currSubCtrl).getTestElement();
+            } else if (currSubCtrl instanceof TestElement) {
+                nextSubController = (TestElement) currSubCtrl;
+            }
+            
+            if (!(nextSubController.isEnabled()) && (nextSubController instanceof Controller || nextSubController instanceof Sampler)) {
+                setNextSubController();
+            }
+        } else {
+            nextSubController = null;
+        }
+    }
+}
+
+/**
+ * This abstraction is necessary for unit tests, in which case a deterministic
+ * generator replaces the random one
```

Comment:
Why can't seed be used in order to make test consistent?
There are at least two ways:
1) Hard-code seed value right into the test
2) Use random seed, and enable configuring the seed from the command line. That way if a test with some specific seed value fails, we can reproduce it by supplying that particular value.